### PR TITLE
Fix documentation on Euler angle convention

### PR DESCRIPTION
**Pull request by azeey**
# 🦟 Bug fix

## Summary

The convention stated in the pose.sdf spec was different from the one documented in http://sdformat.org/tutorials?tut=specify_pose.

\cc @FirefoxMetzger 

## Checklist
- [x] Signed all commits for DCO
- ~[ ] Added tests~
- [x] Updated documentation (as needed)
- ~[ ] Updated migration guide (as needed)~
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: sdf/1.9/pose.sdf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <!-- Pose -->
 <element name="pose" type="pose" default="0 0 0   0 0 0" required="0">
   <description>
-    A pose (x, y, z, r, p, y) expressed in the frame named by @relative_to. The first three components (x, y, z) represent the position of the element's origin (in the @relative_to frame) and the last three components (roll, pitch, yaw) represent the orientation of the element as a sequence of intrinsic Euler rotations.
+    A pose (x, y, z, r, p, y) expressed in the frame named by @relative_to. The first three components (x, y, z) represent the position of the element's origin (in the @relative_to frame) and the last three components (roll, pitch, yaw) represent the orientation of the element as a sequence of extrinsic X-Y-Z Euler rotations. See https://sdformat.org/tutorials?tut=specify_pose.
```
Comment (scpeters) [inline sdf/1.9/pose.sdf:4]: https seems to be broken on sdformat.org at the moment, but otherwise this looks good

**Comment by FirefoxMetzger** [comment]
Thanks for the ping @azeey ; I wasn't aware of this ... and it could explain a bug that I've been chasing for the past few days.

That said, there is something funny about the [tutorial document](http://sdformat.org/tutorials?tut=specify_pose) regarding the nomenclature of extrinsic/intrinsic and the matrix being shown. The way I remember this is that _intrinsic_ rotations are given in the order in which the matrices are multiplied (with a column vector), and _extrinsic_ rotations are given in the order in which a (western) person would write them down (left-to-right). [This implies - as it should - that any intrinsic rotation can be written as an extrinsic one by swapping the order and vice versa.]

Assuming what I said above makes sense, then the equation shown in the tutorial should be an *extrinsic* `Yaw Pitch Roll` (`ZYX`) rotation or an _intrinsic_ `roll pitch yaw` (`xyz`) rotation. This would be contradicting the text, which says that it is an _extrinsic_ `Roll Pitch Yaw` (`XYZ`) rotation.

Tracing the code, Pose3 [builds a quaternion from the given Euler angles](https://github.com/ignitionrobotics/ign-math/blob/c3ab4dc33417d94e25f3ec91cf2b99c1b38196eb/include/ignition/math/Pose3.hh#L117) which is defined in ign-math:
https://github.com/ignitionrobotics/ign-math/blob/ign-math6/include/ignition/math/Quaternion.hh#L322-L340

I found a very similar looking set of equations to the above on Wikipedia: [Euler angles to quaternion conversion](https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles#Euler_angles_to_quaternion_conversion), which are given for a Body 3-2-1 rotation. The section above the linked one explains that a Body 3-2-1 rotation means an _intrinsic_ `xyz` Euler rotation (_extrinsic_ `ZYX` Euler).

This makes me believe that the equations in the tutorial are correct, but that the text description is wrong. That said, it's easy to confuse these rotations, so it would be good if someone else can confirm this.

**Comment by azeey** [comment]
> The way I remember this is that intrinsic rotations are given in the order in which the matrices are multiplied (with a column vector), and extrinsic rotations are given in the order in which a (western) person would write them down (left-to-right). [This implies - as it should - that any intrinsic rotation can be written as an extrinsic one by swapping the order and vice versa.]

I believe the core meaning of intrinsic vs extrinsic is whether the sequence of rotation happens in the body frame (intrinsic) or the world frame (extrinsic). So, for me, when I read `extrinsic XYZ` or `extrinsic RPY`, I interpret it as rotation in the world's X-axis, followed by rotation in the world's Y-axis, and finally rotation in the world's Z-axis. 

> I found a very similar looking set of equations to the above on Wikipedia: Euler angles to quaternion conversion, which are given for a Body 3-2-1 rotation. The section above the linked one explains that a Body 3-2-1 rotation means an intrinsic xyz Euler rotation (extrinsic ZYX Euler).

I couldn't find the  section that says this is an intrinsic xyz. The linked article also seems to use the left-hand coordinates. SDFormat uses right-hand coordinates, so maybe that's missing peace of information?

**Comment by FirefoxMetzger** [comment]
> I believe the core meaning of intrinsic vs extrinsic is whether the sequence of rotation happens in the body frame (intrinsic) or the world frame (extrinsic). So, for me, when I read extrinsic XYZ or extrinsic RPY, I interpret it as rotation in the world's X-axis, followed by rotation in the world's Y-axis, and finally rotation in the world's Z-axis.

Yep, on this part we agree. The part that I was questioning (perhaps because I am confusing myself) is the step afterward: How does one go from `extrinsic XYZ`/`extrinsic RPY` to the sequence of matrix multiplications shown in the tutorial?

```
R_px = R_z(yaw)*R_y(pitch)*R_x(roll)
```

In the above rotation chain, the simple rotation `R_x(roll)` (which rotates `roll` radians around the current frame's x-axis) acts first. It takes a vector expressed in the initial frame (frame0) and outputs a vector expressed a the rotated frame (frame1). Here, the rotation happens around frame0's x-axis. Next, `R_y(pitch)` acts. It gets as input the vector expressed in frame1. Around which frame's y-axis does `R_y(pitch)` rotate? (I think around frame1's y-axis; hence the rotation being `intrinsic xyz` or `extrinsic ZYX`)

Perhaps [this section](https://en.wikipedia.org/wiki/Davenport_chained_rotations#Conversion_between_intrinsic_and_extrinsic_rotations) (also from Wikipedia) which uses `intrinsic xyz`/`extrinsic ZYX` in its example is better able to describe my perspective?

> I couldn't find the section that says this is an intrinsic xyz. The linked article also seems to use the left-hand coordinates. SDFormat uses right-hand coordinates, so maybe that's missing peace of information?

Good catch @azeey; I totally missed that (and have - so far - never seen people use left-hand frames on purpose). Interestingly, all the figures in the article use right-hand frames 😄 Goes to show how inhomogeneous definitions are. I spent some time digging up the source of that article (which is actually a [NASA paper from the 70's](https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf)), and they, too, use right-hand frames ... 

The piece that throws me in the NASA paper is that they define a transformation from the body frame to the lab frame (in our lingo: child -> parent); this adds another level of inversion (we want parent -> child). So essentially, their intrinsic rotations are our extrinsic rotations and vice versa. In response, Wikipedia takes the only logical step in their article: they sometimes use parent->child and sometimes use child -> parent (╯°□°)╯︵ ┻━┻. 

Maybe it is easier to leave this article alone, forget that I linked it, and look for a different source of truth; @azeey do you have something authoritative regarding the conversion from Euler angles to quaternions?

**Review by aaronchongth — APPROVED** [review]
Thanks for catching and fixing that mistake! LGTM

**Comment by traversaro** [comment]
> Maybe it is easier to leave this article alone, forget that I linked it, and look for a different source of truth; @azeey do you have something authoritative regarding the conversion from Euler angles to quaternions?

In my limited experience, when documenting a given use of Euler angles it is better to avoid at all to use "extrinsic" or "intrisic" nomenclature, exactly to avoid all the confusion described in this issue. The least ambigous thing to do to me seems to just remove any reference to "extrinsic" or "intristic" or any external definition, and just leave the link to http://sdformat.org/tutorials?tut=specify_pose as the only single source of true for the semantics of SDF's euler angles.

**Comment by azeey** [comment]
> Maybe it is easier to leave this article alone, forget that I linked it, and look for a different source of truth; @azeey do you have something authoritative regarding the conversion from Euler angles to quaternions?

Appendix B.2 of [Modern Robotics](http://hades.mech.northwestern.edu/index.php/Modern_Robotics) discusses RPY angles, but it doesn't say "extrinsic" or "intrinsic", just "with respect to the fixed frame" or "with respect to the body frame".

> Perhaps this section (also from Wikipedia) which uses intrinsic xyz/extrinsic ZYX in its example is better able to describe my perspective?

The wikipedia article says
>Any extrinsic rotation is equivalent to an intrinsic rotation by the same angles but with inverted order of elemental rotations, and vice versa. For instance, the intrinsic rotations x-y’-z″ by angles α, β, γ are equivalent to the extrinsic rotations z-y-x by angles γ, β, α. Both are represented by a matrix
> image

I agree that's an extrinsic z-y-x. And in our case, the formula is 
image
 which would make it an extrinsic x-y-z.

> The least ambigous thing to do to me seems to just remove any reference to "extrinsic" or "intristic" or any external definition, and just leave the link to http://sdformat.org/tutorials?tut=specify_pose as the only single source of true for the semantics of SDF's euler angles.

I agree with this. We'll have to update the website since it also states "extrinsic".

**Review by scpeters — APPROVED** [review]

**Pull request merged**